flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -350,42 +355,105 @@ private boolean isPartialResourceConfigured(JobGraph jobGraph) {
 
 	private CompletableFuture<Void> persistAndRunJob(JobGraph jobGraph) throws Exception {
 		jobGraphWriter.putJobGraph(jobGraph);
+		return runJob(jobGraph, false);
+	}
+
+	// TODO remove future
+	// same error in recovery -- kill process
+	private CompletableFuture<Void> runJob(JobGraph jobGraph, boolean isRecovery) {
+		Preconditions.checkState(!runningJobs.containsKey(jobGraph.getJobID()));
+		long jobManagerInitializationStarted = System.currentTimeMillis();
+
+		CompletableFuture<JobManagerRunner> jobManagerRunnerFuture = createJobManagerRunner(jobGraph)
+			.thenApplyAsync(
+				FunctionUtils.uncheckedFunction(jobManagerRunner -> {
+					jobManagerRunner.start();
+					return jobManagerRunner;
+				}),
+			getRpcService().getExecutor()); // execute in separate pool to avoid blocking the Dispatcher
```

Comment:
I'd make this part of the `createJobManagerRunner` call so that `jobManagerRunnerFuture` will complete with a running `JobManagerRunner`.